cs/src/core/Device/AzurePageBlobDevice.cs

```diff
@@ -0,0 +1,136 @@
+        /// <param name="deleteOnClose">
+        /// True if the program should delete all blobs created on call to <see cref="Close">Close</see>. False otherwise. 
+        /// The container is not deleted even if it was created in this constructor
+        /// </param>
+        public AzurePageBlobDevice(string connectionString, string containerName, string blobName, bool deleteOnClose = false)
+            : base(connectionString + "/" + containerName + "/" + blobName, PAGE_BLOB_SECTOR_SIZE)
+        {
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(connectionString);
+            CloudBlobClient client = storageAccount.CreateCloudBlobClient();
+            container = client.GetContainerReference(containerName);
+            container.CreateIfNotExists();
+            blobs = new ConcurrentDictionary<int, CloudPageBlob>();
+            this.blobName = blobName;
+            this.deleteOnClose = deleteOnClose;
+        }
+
+        /// <summary>
+        /// <see cref="IDevice.Close">Inherited</see>
+        /// </summary>
+        public override void Close()
+        {
+            // Unlike in LocalStorageDevice, we explicitly remove all page blobs if the deleteOnClose flag is set, instead of relying on the operating system
+            // to delete files after the end of our process. This leads to potential problems if multiple instances are sharing the same underlying page blobs.
+            //
+            // Since this flag is presumably only used for testing though, it is probably fine.
+            if (deleteOnClose)
+            {
+                foreach (var entry in blobs)
+                {
+                    entry.Value.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.Close">Inherited</see>
+        /// </summary>
+        public override void DeleteSegmentRange(int fromSegment, int toSegment)
+        {
+            for (int i = fromSegment; i < toSegment; i++)
+            {
+                if (blobs.TryRemove(i, out CloudPageBlob blob))
+                {
+                    blob.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.ReadAsync(int, ulong, IntPtr, uint, IOCompletionCallback, IAsyncResult)">Inherited</see>
+        /// </summary>
+        public override unsafe void ReadAsync(int segmentId, ulong sourceAddress, IntPtr destinationAddress, uint readLength, IOCompletionCallback callback, IAsyncResult asyncResult)
+        {
+            CloudPageBlob pageBlob = GetOrAddPageBlob(segmentId);
+
+            // Even though Azure Page Blob does not make use of Overlapped, we populate one to conform to the callback API
+            Overlapped ov = new Overlapped(0, 0, IntPtr.Zero, asyncResult);
+            NativeOverlapped* ovNative = ov.UnsafePack(callback, IntPtr.Zero);
```

Comment:
this will leak unless the callback is actually called (e.g., due to exceptions, etc.). Probably need to wrap with try/finally.

Reply:
I have been unable to find any documentation on what exceptions are thrown from the Azure call. However, if one is thrown, it will be unhandled and the application will be halted, and memory leak in this situation is a non-issue.

Throwing an exception seems reasonable because there is little the code can do to recover at the Device level.

Reply:
Unfortunately unhandled exceptions from thread pool threads do not crash the app, and instead terminate the background thread and fail silently. It is also a common pattern to retry a failed operation, especially for cloud operations that suffer from intermittent failures. But aside from memory leak, I was more worried about the blob leak - I want to make sure we don't create blobs in the remote storage that are never cleaned up.

---
In reply to: [303554075](https://github.com/microsoft/FASTER/pull/147#discussion_r303554075) [](ancestors = 303554075)